DFE-Digital/sts-plan-technology-for-your-school
# Chore: Improvements to the dbo schema

## Overview

Addresses ticket [#231142](https://dfe-ssp.visualstudio.com/s190-schools-technology-services/_workitems/edit/231142)

Various improvements to the dbo schema to improve efficiency and usability

These improvements are all outcomes of a spike on SQL improvements documented in [This ADR](https://github.com/DFE-Digital/sts-plan-technology-for-your-school/blob/development/docs/adr/0040-sql-schema-improvements.md)

## Changes

### Minor
- Change `GetCurrentSubmissionId` to a UDF as these are good for returning a single value
- Simplify uses of `GetCurrentSubmissionId` with the new UDF
- Reduce duplication of data in the `SubmitAnswer` stored proc
- Remove existing duplicated data 
- Add an index to `dbo.response` on submissionId
- Fix the SeedTestData project to support caching

### Non-functional changes (e.g. documentation)
- The ADR had one recommendation in the wrong section, moved that and updated it to approved as that was missed at the time of merging it

## How to review the PR

- Check that nothing from the ticket/ADR has been missed for the dbo schema
- Functional test:
  - Check out development
  - Run the script to create the seeded test database
  - (Note that you may need to add `services.AddSingleton<IQueryCacher, QueryCacher>();` after creating services in `Program.cs` of the SeedTestData project because this got missed when we added caching, fixed in this PR though)
  - Point plan tech at it
  - Complete some topics
  - Complete them again
  - Note the duplication in the `dbo.question` and `dbo.answers` tables
  - Run this new script
  - See that the duplication has been removed, but that the `dbo.response` table is still in order connected to the correct question and answer text
  - Check that "Last updated" hasn't changed on the self assessment page
  - Everything should continue working as before

## Checklist
